<a href="https://colab.research.google.com/github/agabrielabarros/Programando_em_Python/blob/main/Conhecendo_bibliotecas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conhecendo Bibliotecas

In [ ]:
import pandas as pd

In [ ]:
dados_churn = pd.read_json("/content/drive/MyDrive/Colab Notebooks/Limpeza de dados Json/dataset-telecon.json")

In [ ]:
dados_churn.head()

,id_cliente,Churn,cliente,telefone,internet,conta
0,0002-ORFBO,nao,"{'genero': 'feminino', 'idoso': 0, 'parceiro':...","{'servico_telefone': 'sim', 'varias_linhas': '...","{'servico_internet': 'DSL', 'seguranca_online'...","{'contrato': None, 'faturamente_eletronico': N..."
1,0003-MKNFE,nao,"{'genero': 'masculino', 'idoso': 0, 'parceiro'...","{'servico_telefone': 'sim', 'varias_linhas': '...","{'servico_internet': 'DSL', 'seguranca_online'...","{'contrato': 'mes a mes', 'faturamente_eletron..."
2,0004-TLHLJ,sim,"{'genero': 'masculino', 'idoso': 0, 'parceiro'...","{'servico_telefone': 'sim', 'varias_linhas': '...","{'servico_internet': 'fibra otica', 'seguranca...","{'contrato': 'mes a mes', 'faturamente_eletron..."
3,0011-IGKFF,sim,"{'genero': 'masculino', 'idoso': 1, 'parceiro'...","{'servico_telefone': 'sim', 'varias_linhas': '...","{'servico_internet': 'fibra otica', 'seguranca...","{'contrato': 'mes a mes', 'faturamente_eletron..."
4,0013-EXCHZ,sim,"{'genero': 'feminino', 'idoso': 1, 'parceiro':...","{'servico_telefone': 'sim', 'varias_linhas': '...","{'servico_internet': 'fibra otica', 'seguranca...","{'contrato': 'mes a mes', 'faturamente_eletron..."


In [ ]:
dados_churn['conta'][0]

{'contrato': None,
 'faturamente_eletronico': None,
 'metodo_pagamento': None,
 'cobranca': {'mensal': None, 'Total': None}}

O método json_normalize() é uma função da biblioteca Pandas utilizada para transformar dados em formato JSON em um DataFrame. Ela é muito útil quando se trabalha com dados em JSON, pois muitas vezes eles estão em formato aninhado e precisam ser normalizados para que possam ser trabalhados facilmente.

> Vamos deixar essa coluna de 'contas' em um formato normal com normalize. Cada chave dessa veio como colunas, de forma separada

In [ ]:
pd.json_normalize(dados_churn['conta']).head()

,contrato,faturamente_eletronico,metodo_pagamento,cobranca.mensal,cobranca.Total
0,None,None,None,NaN,None
1,mes a mes,nao,cheque pelo correio,59.9,542.4
2,mes a mes,sim,cheque eletronico,73.9,280.85
3,mes a mes,sim,cheque eletronico,98.0,1237.85
4,mes a mes,sim,cheque pelo correio,83.9,267.4


> Vamos fazer isso para com mais uma coluna que estão com os dados juntos.

In [ ]:
pd.json_normalize(dados_churn['telefone']).head()

,servico_telefone,varias_linhas
0,sim,nao
1,sim,sim
2,sim,nao
3,sim,nao
4,sim,nao


Mas fazer isso uma por uma dá muito trabalho... Se quisermos um dataframe totalmente normalizado em uma tabela que possamos trabalhar é possível fazer tudo de uma vez só, ao invés de uma a uma.

## Transformando dados em uma tabela

> Podemos aplicar o método json_normalize() em um dataframe (como fizemos acima) ou em um objeto JSON. A principal diferença é que, ao normalizar uma coluna específica, aplicamos o método no dataframe. Já ao normalizar todo o JSON de uma vez, aplicamos o método no objeto JSON.

* Em um DataFrame ⇾ normaliza apenas uma coluna

* Em um objeto JSON ⇾ normaliza todas as colunas aninhadas de uma só vez

In [ ]:
import json

In [ ]:
with open('/content/drive/MyDrive/Colab Notebooks/Limpeza de dados Json/dataset-telecon.json') as f:
  json_bruto = json.load(f)

In [ ]:
json_bruto

In [ ]:
dados_normalizados = pd.json_normalize(json_bruto)
dados_normalizados.head()

,id_cliente,Churn,cliente.genero,cliente.idoso,cliente.parceiro,cliente.dependentes,cliente.tempo_servico,telefone.servico_telefone,telefone.varias_linhas,internet.servico_internet,...,internet.backup_online,internet.protecao_dispositivo,internet.suporte_tecnico,internet.tv_streaming,internet.filmes_streaming,conta.contrato,conta.faturamente_eletronico,conta.metodo_pagamento,conta.cobranca.mensal,conta.cobranca.Total
0,0002-ORFBO,nao,feminino,0,sim,sim,9.0,sim,nao,DSL,...,sim,nao,sim,sim,nao,None,None,None,NaN,None
1,0003-MKNFE,nao,masculino,0,nao,nao,9.0,sim,sim,DSL,...,nao,nao,nao,nao,sim,mes a mes,nao,cheque pelo correio,59.9,542.4
2,0004-TLHLJ,sim,masculino,0,nao,nao,4.0,sim,nao,fibra otica,...,nao,sim,nao,nao,nao,mes a mes,sim,cheque eletronico,73.9,280.85
3,0011-IGKFF,sim,masculino,1,sim,nao,13.0,sim,nao,fibra otica,...,sim,sim,nao,sim,sim,mes a mes,sim,cheque eletronico,98.0,1237.85
4,0013-EXCHZ,sim,feminino,1,sim,nao,3.0,sim,nao,fibra otica,...,nao,nao,sim,sim,nao,mes a mes,sim,cheque pelo correio,83.9,267.4


Exatamente! O **with** é utilizado para abrir o arquivo JSON e carregar o conteúdo de forma segura e eficiente, garantindo que o arquivo será fechado após o uso. Abrimos o arquivo “dados.json” e usamos o método *json.load()*. O método **json.load()** é uma função Python que permite carregar um arquivo JSON em memória e convertê-lo em um objeto Python que pode ser usado no programa. Logo após, inserimos o arquivo carregado no método **json_normalize** que converte o arquivo JSON em uma tabela relacional, em que as chaves do objeto JSON são transformadas em colunas e seus valores são adicionados nas linhas correspondentes.

Agora sim nosso dataframe está pronto para começarmos a **tratar**.

# Trabsformaçõs iniciais dos dados

## Conhecendo as informações

A base de dados contém colunas além do ID dos clientes e o churn:

<b>Cliente:</b>

* `genero`: gênero (masculino e feminino)
* `idoso`: informação sobre um cliente ter ou não idade igual ou maior que 65 anos
* `parceiro`: se o cliente possui ou não um parceiro ou parceira
* `dependentes`: se o cliente possui ou não dependentes
* `tempo_servico`: meses de contrato do cliente

<b>Serviço de telefonia</b>

 * `servico_telefone`: assinatura de serviço telefônico
 * `varias_linhas`: assinatura de mais de uma linha de telefone


<b>Serviço de internet</b>


 * `servico_internet`: assinatura de um provedor internet
 * `seguranca_online`: assinatura adicional de segurança online
 * `backup_online`: assinatura adicional de backup online
 * `protecao_dispositivo`: assinatura adicional de proteção no dispositivo
 * `suporte_tecnico`: assinatura adicional de suporte técnico, menos tempo de espera
 * `tv_streaming`: assinatura de TV a cabo
 * `filmes_streaming`: assinatura de streaming de filmes


<b>Conta</b>

 * `contrato`: tipo de contrato
 * `faturamente_eletronico`: se o cliente prefere receber online a fatura
 * `metodo_pagamento`: forma de pagamento
 * `cobranca.mensal`: total de todos os serviços do cliente por mês
 * `cobranca.Total`: total gasto pelo cliente

In [ ]:
dados_normalizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7344 entries, 0 to 7343
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id_cliente                     7344 non-null   object 
 1   Churn                          7344 non-null   object 
 2   cliente.genero                 7344 non-null   object 
 3   cliente.idoso                  7344 non-null   int64  
 4   cliente.parceiro               7344 non-null   object 
 5   cliente.dependentes            7344 non-null   object 
 6   cliente.tempo_servico          7336 non-null   float64
 7   telefone.servico_telefone      7344 non-null   object 
 8   telefone.varias_linhas         7344 non-null   object 
 9   internet.servico_internet      7344 non-null   object 
 10  internet.seguranca_online      7344 non-null   object 
 11  internet.backup_online         7344 non-null   object 
 12  internet.protecao_dispositivo  7344 non-null   o

Temos 7344 entradas que vão de 0 à 7343. Esta informação nos diz que não há nenhum índice saltado.

Se tivéssemos 7344 entradas que vão de 0 à 7500, por exemplo, teríamos um problema de índices na base de dados.

Além disso, temos o total de 21 colunas.

Na tabela propriamente dita, vemos quatro colunas com os seguintes dados:

* #, que possui o índice
* Column, com o nome da nossa coluna
* Non-Null Count, com a quantidade de valores não nulos
* DType, com o tipo do dado

Em **Non-Null Count**, percebemos que a maioria dos valores correspondem a 7344 (o número de amostras que temos). Contudo, alguns dados contém outros valores como, por exemplo, a conta.contrato que possui 7312 amostras não nulas.

Isso significa que há dados nulos nesse conjunto. Posteriormente, veremos como trabalhar com ele.

Em Dtype, a maioria dos valores correspondem a object, o que significa que contém strings. Contudo, alguns possuem valores diferentes como, por exemplo, a cliente.idoso que possui o tipo int64 (somente valores inteiros).

Se subirmos o dicionário e acessarmos o retorno do dataframe, veremos que a coluna cliente.idoso possui somente os valores binários 0 e 1. Com isso, concluiremos que o tipo int64, faz sentido para estes dados.

Voltando à tabela inferior, temos a cliente.tempo_servico, que possui valores do tipo float64. Faz sentido ser um valor numérico, pois esta coluna trata da quantidade de meses que uma pessoa cliente possui o serviço ativo.

A coluna conta.cobranca.mensal também possui o tipo de dado float64. Dado que se trata de um valor monetário (em reais), esta tipagem também faz sentido.

Já a colunaconta.cobranca.total, que também trabalha com valores monetários, possui o tipo object, que não faz sentido para moedas. Ela deveria ser do tipo float64 ou, na pior hipótese, um int64.

O que devemos fazer quando temos colunas com tipos incorretos no banco de dados — principalmente quando pretendemos inserir estes dados em um modelo de Machine Learning? Transformá-los na tipagem correta por meio de um cast, permitindo que este modelo possa tratá-los de forma correta e gerar resultados assertivos.

## Modificando o tipo da coluna

Identificando valores estranhos
Para realizar o cast, acessaremos a célula vazia localizada embaixo da tabela que foi retornada por nosso último comando.

Nela, escreveremos o nome do campo de dados (dados_normalizados[]), que terá entre colchetes e aspas simples o nome da coluna a ser corrigida (conta.cobranca.Total).

À direita dos colchetes, adicionaremos o sinal de igual e a operação que queremos realizar — neste caso, queremos retornar a mesma coluna, só que com o tipo diferente, portanto copiaremos todo o conteúdo à esquerda do sinal de igual e colaremos à sua direita, adicionando o método .astype(), informando entre parênteses o novo tipo de dado: float.

In [ ]:
dados_normalizados['conta.cobranca.Total'] = dados_normalizados['conta.cobranca.Total'].astype(float)

ValueError: could not convert string to float: ' '

Executaremos a célula com "Shift+Enter", e abaixo dela surgirá o retorno de um erro. Ele informa que não podemos converter a string da coluna para o tipofloat.

Isto significa que temos uma string estranha nesta coluna. Por este motivo, ela adquiriu o tipo object.

Temos que tratar este valor estranho da coluna antes de realizar o cast. Faremos isso a seguir.

## Identificando e tratando strings vazias